resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -110,65 +108,103 @@ private[spark] class Client(
+                      .withNewTargetPort(DRIVER_LAUNCHER_SERVICE_INTERNAL_PORT)
+                      .build()
+                    val service = kubernetesClient.services().createNew()
+                      .withNewMetadata()
+                        .withName(kubernetesAppId)
+                        .endMetadata()
+                      .withNewSpec()
+                        .withType("NodePort")
+                        .withSelector(selectors)
+                        .withPorts(driverLauncherServicePort)
+                        .endSpec()
+                      .done()
+                    try {
+                      sparkConf.set("spark.kubernetes.driver.service.name",
+                        service.getMetadata.getName)
+                      sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
+                      sparkConf.setIfMissing("spark.blockmanager.port",
+                        DEFAULT_BLOCKMANAGER_PORT.toString)
+                      val driverLauncher = getDriverLauncherClient(kubernetesClient, service)
+                      val ping = Retry.retry(5, 5.seconds) {
+                        driverLauncher.ping()
+                      }
+                      ping onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val submitComplete = ping.flatMap { _ =>
+                        Future {
+                          sparkConf.set("spark.driver.host", pod.getStatus.getPodIP)
+                          val submitRequest = buildSubmissionRequest()
+                          driverLauncher.create(submitRequest)
                         }
-                    }
-                    val submitComplete = ping andThen {
-                      case Success(_) =>
-                        driverLauncher.create(submitRequest)
-                        submitCompletedFuture.set(true)
-                    }
-                    submitComplete onFailure {
-                      case t: Throwable =>
-                        if (!submitCompletedFuture.isDone) {
+                      }
+                      submitComplete onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val adjustServicePort = submitComplete.flatMap { _ =>
+                        Future {
+                          // After submitting, adjust the service to only expose the Spark UI
+                          val uiServicePort = new ServicePortBuilder()
+                            .withName(UI_PORT_NAME)
+                            .withPort(uiPort)
+                            .withNewTargetPort(uiPort)
+                            .build
```

Comment:
It's typical Spark style to always include parenthesis when calling a function
.build -> .build(), same else where